tensorflow/core/kernels/depthwise_conv_grad_op.cc

```diff
@@ -1139,10 +1145,11 @@ class DepthwiseConv2dNativeBackpropFilterOp : public OpKernel {
     bool use_cudnn = std::is_same<Device, GPUDevice>::value &&
                      (in_depth == 1 ||
                       (use_cudnn_grouped_conv_ &&
-                       IsCudnnSupportedFilterSize(/*filter_rows=*/filter_rows,
-                                                  /*filter_cols=*/filter_cols,
-                                                  /*in_depth=*/in_depth,
-                                                  /*out_depth=*/out_depth)));
+                       (IsCudnnSupportedFilterSize(/*filter_rows=*/filter_rows,
+                                                   /*filter_cols=*/filter_cols,
+                                                   /*in_depth=*/in_depth,
+                                                   /*out_depth=*/out_depth) ||
+                        OpDeterminismRequired())));
```

Comment:
If determinism is enabled, we use cudnn even if `IsCudnnSupportedFilterSize` returns false. Does this mean `IsCudnnSupportedFilterSize` returns false even if cudnn is supported?

The declaration of `IsCudnnSupportedFilterSize` has a comment saying to see the [cuDNN 7.6.3 release notes](https://docs.nvidia.com/deeplearning/cudnn/release-notes/rel_7xx.html#rel_763). The release notes state "Improved grouped convolution for `cudnnConvolutionBackwardFilter()` in the configuration below", then lists the configuration that `IsCudnnSupportedFilterSize` matches. However, the relase notes use the word "improved", not "added support for", so I'm assuming cuDNN is supported in more cases . Additionally, the cuDNN documentation does not state that cuDNN is only supported in certain cases. So it does seem cuDNN is supported in all cases.

If this is true, perhaps `IsCudnnSupportedFilterSize` should be renamed to `ShouldCudnnBeUsedFilterSize`.

Reply:
The cuDNN developer guide section on [grouped convolution](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#grouped-convolutions) states that, "Grouped convolutions are supported for all formats that are currently supported by the functions `cudnnConvolutionForward()`, `cudnnConvolutionBackwardData()` and `cudnnConvolutionBackwardFilter()`."

Yes, it appears to me that TensorFlow currently restricts its use of cuDNN grouped convolution only for performance advantages, not because of missing functionality in cuDNN. I have wondered if the restrictions in TensorFlow are no longer needed, or perhaps were never needed (having arisen from confusion/misunderstanding due to ambiguities in the release notes).

In versions of cuDNN earlier than 7.6.3, there were functional issues, which appear to have been resolved.

For example, [version 7.5.0](https://docs.nvidia.com/deeplearning/cudnn/release-notes/rel_7xx.html#rel_750) **_fixed_** "Some Grouped Convolutions (particularly those used in Depthwise-Separable convolutions) may return INTERNAL_ERROR if they have all inputs/outputs as NHWC-packed and do not match one of the following criteria: ..."

Another example is that in [version 7.2.1](https://docs.nvidia.com/deeplearning/cudnn/release-notes/rel_7xx.html#rel_721) contained the limitation that "Grouped convolutions are not supported in the TRUE_HALF_CONFIG (convDesc is CUDNN_DATA_HALF) data type configuration. As a workaround, the PSEUDO_HALF_CONFIG (convDesc is CUDNN_DATA_FLOAT) data type configuration can be used without losing any precision."

Yes, I agree that `IsCudnnSupportedFilterSize` should be renamed. This aspect of the existing code was confusing to me because of the use of the word "supported." I propose: `ShouldCudnnGroupedConvolutionFilterSizeBeUsed`. I can submit a follow-up PR for this, if you agree.

Reply:
Also, the only reported functional limitation on grouped convolution in cuDNN [version 8.x](https://docs.nvidia.com/deeplearning/cudnn/release-notes/rel_8.html#rel_8) (persisting into version 8.4.0) is: "For the `_ALGO_0` algorithm of convolution backward data and backward filter, grouped convolution with groups larger than 1 and with odd product of dimensions `C`, `D` (if 3D convolution), `H`, and `W` is not supported on devices older than Volta. To prevent a potential illegal memory access by an instruction that only has a 16-bit version in Volta and above, pad at least one of the dimensions to an even value."

I'm confused about what "groups larger than 1" specifically refers to (and I'm seeking internal clarification on this), but for the purposes of deterministic functionality, the `_ALGO_0` algorithms are, luckily, [nondeterministic](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#reproducibility). Therefore, this limitation does not impact this current PR to add deterministic functionality.

Reply:
I think `ShouldCudnnGroupedConvolutionFilterSizeBeUsed` sounds good. Maybe remove `FilterSize` since the function name is pretty long.

Perhaps we should investigate always using cuDNN, but we can always do that later and it has nothing to do with determinism. Presumably any limitations on `_ALGO_0` won't matter for correctness, since autotune will just choose another algorithm.

Reply:
I think it's a good idea to investigate (beyond this PR) always using cuDNN grouped convolutions for TensorFlow depthwise convolutions; I wonder if it would increase performance overall. Right now, cuDNN is used in a very restricted way: only half-precision, only square filters of size 1, 3, 5, or 7 on a side, and only when `channel_multiplier` is 1. Good point about auto-tuning elegantly handling the `_ALGO_0` limitation on pre-Volta GPUs.

Reply:
I'd like to also bring attention to [this comment](https://github.com/tensorflow/tensorflow/blob/fa2b6b881d12d94b8b9a8a8222d27a1bd6824dbd/tensorflow/core/kernels/depthwise_conv_op.cc#L418-L420) from @chsigg (added 2018-04-30, which, for context, was between the release of 7.1.3 and 7.1.4 of cuDNN):

```
// TODO(csigg): Have autotune decide if native is faster than cuDNN.
// If in_depth==1, this operation is just a standard convolution.
// Depthwise convolution is a special case of cuDNN's grouped convolution.
```